File: geocoding.py
"""
Geocoding support for converting place names to coordinates.

Uses Nominatim (OpenStreetMap) geocoding service which is free
and doesn't require an API key.
"""
import requests
import time
from threading import Lock
from typing import Optional, Dict, Any


class Geocoder:
    """Simple geocoder using Nominatim API."""

    BASE_URL = "https://nominatim.openstreetmap.org"
    MAX_PLACE_LENGTH = 200

    def __init__(self, timeout: int = 10):
        """
        Initialize the geocoder.

        Args:
            timeout: Request timeout in seconds (default: 10)
        """
        self.timeout = timeout
        self.session = requests.Session()
        # Nominatim requires a User-Agent per their usage policy
        self.session.headers.update({
            'User-Agent': 'IFO-CLI/1.0 (Aircraft tracking tool)'
        })
        # Rate limiting to comply with Nominatim usage policy (max 1 req/sec)
        self._last_request_time = 0.0
        self._lock = Lock()

    def geocode(self, place: str) -> Optional[Dict[str, Any]]:
        """
        Convert a place name to coordinates.

        Args:
            place: Place name (e.g., "San Francisco", "London, UK")

        Returns:
            Dictionary with 'lat', 'lon', and 'display_name' if found,
            None if not found

        Raises:
            requests.RequestException: If the geocoding request fails
            ValueError: If place name is invalid
        """
        if not place or not place.strip():
            raise ValueError("Place name cannot be empty")

        if len(place) > self.MAX_PLACE_LENGTH:
            raise ValueError(f"Place name too long (max {self.MAX_PLACE_LENGTH} characters)")

        # Rate limiting: ensure at least 1 second between requests (Nominatim policy)
        with self._lock:
            time_since_last = time.time() - self._last_request_time
            if time_since_last < 1.0:
                time.sleep(1.0 - time_since_last)
            self._last_request_time = time.time()

        params = {
            'q': place.strip(),
            'format': 'json',
            'limit': 1
        }

        response = self.session.get(
            f"{self.BASE_URL}/search",
            params=params,
            timeout=self.timeout
        )
        response.raise_for_status()

        data = response.json()

        if not data or len(data) == 0:
            return None

        result = data[0]

        # Validate response structure
        if 'lat' not in result or 'lon' not in result or 'display_name' not in result:
            raise ValueError("Unexpected response format from geocoding service")

        return {
            'lat': float(result['lat']),
            'lon': float(result['lon']),
            'display_name': result['display_name']
        }

    def close(self):
        """Close the HTTP session."""
        self.session.close()

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()

File: src/geocoding.rs
//! Geocoding module for converting place names to coordinates.
//!
//! Uses Nominatim (OpenStreetMap) geocoding service which is free
//! and doesn't require an API key.

use governor::{Jitter, Quota, RateLimiter};
use nonzero::nonzero;
use reqwest::Client;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::Mutex;

use crate::error::{IfoError, Result};
use crate::models::{Location, NominatimResult};

const MAX_PLACE_LENGTH: usize = 200;

/// Geocoder using Nominatim API with rate limiting.
pub struct Geocoder {
    client: Client,
    base_url: String,
    rate_limiter: Arc<
        Mutex<
            RateLimiter<
                governor::state::NotKeyed,
                governor::state::InMemoryState,
                governor::clock::DefaultClock,
            >,
        >,
    >,
}

impl Geocoder {
    /// Create a new geocoder with rate limiting.
    pub fn new(timeout_secs: u64) -> Result<Self> {
        let client = Client::builder()
            .user_agent("IFO-CLI/2.0 (Aircraft tracking tool)")
            .timeout(Duration::from_secs(timeout_secs))
            .build()
            .map_err(IfoError::NetworkError)?;

        // Create rate limiter: 1 request per second (Nominatim policy)
        let quota = Quota::per_second(nonzero!(1u32));
        let rate_limiter = RateLimiter::direct(quota);

        Ok(Self {
            client,
            base_url: "https://nominatim.openstreetmap.org".to_string(),
            rate_limiter: Arc::new(Mutex::new(rate_limiter)),
        })
    }

    /// Convert a place name to coordinates.
    pub async fn geocode(&self, place: &str) -> Result<Option<Location>> {
        // Validate input
        let place = place.trim();
        if place.is_empty() {
            return Err(IfoError::EmptyPlaceName);
        }
        if place.len() > MAX_PLACE_LENGTH {
            return Err(IfoError::PlaceNameTooLong {
                max: MAX_PLACE_LENGTH,
            });
        }

        // Rate limiting: wait for permission
        {
            let limiter = self.rate_limiter.lock().await;
            // Add jitter to avoid thundering herd
            limiter
                .until_ready_with_jitter(Jitter::up_to(Duration::from_millis(100)))
                .await;
        }

        // Make request
        let url = format!("{}/search", self.base_url);
        let response = self
            .client
            .get(&url)
            .query(&[("q", place), ("format", "json"), ("limit", "1")])
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            return Err(IfoError::ApiError {
                status: status.as_u16(),
                message: format!("Geocoding failed: {}", text),
            });
        }

        let results: Vec<NominatimResult> = response.json().await?;

        if results.is_empty() {
            return Ok(None);
        }

        let result = &results[0];

        // Parse and validate response
        let lat = result
            .lat
            .parse::<f64>()
            .map_err(|_| IfoError::UnexpectedResponse {
                service: "Nominatim".to_string(),
            })?;
        let lon = result
            .lon
            .parse::<f64>()
            .map_err(|_| IfoError::UnexpectedResponse {
                service: "Nominatim".to_string(),
            })?;

        Ok(Some(Location {
            lat,
            lon,
            display_name: result.display_name.clone(),
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_place_validation() {
        // Test would require async runtime
        // These tests would normally be in tests/geocoding_tests.rs
    }

    #[test]
    fn test_max_place_length() {
        let _long_name = "a".repeat(MAX_PLACE_LENGTH + 1);
        // Would test that this returns an error
    }
}
